### PR TITLE
Correctly URI-encode HTTP Label values

### PR DESCRIPTION
**Pull request by bpholt**
I was a little surprised to find that I didn't need to modify `toSmithy4sHttpUri` to make this test pass, so I wanted to get some feedback before I get too much further into modifying `AwsClient` and `SimpleRestJson` as laid out in #1663.

Needs https://github.com/disneystreaming/alloy/pull/210

### CODE REVIEW

File: modules/aws-http4s/src/smithy4s/aws/internals/AwsSigning.scala [hunk 1]
```diff
@@ -116,77 +126,124 @@ private[aws] object AwsSigning {
...
-              key.toString.toLowerCase + ":" + value.trim
+      preSign(request)
+        .mproduct[String] { case (_, preparedRequest) =>
+          if (disableDoubleEncoding.contains(serviceName))
+            // S3 and S3Control both disable double URL encoding and normalization.
+            preparedRequest.uri.path.toAbsolute.renderString.pure[F]
+          else {
+            // Other services double encode and normalize
+
+            if (preparedRequest.uri.path.segments.exists(s => dotSegments.contains(s.encoded))) {
```
Comment (bpholt) [inline modules/aws-http4s/src/smithy4s/aws/internals/AwsSigning.scala]: FYI, I think it would also be valid to simply use `java.net.URI#normalize` regardless of whether the path contains any dot segments. (The tests pass if I make that change.) I can switch to that if you'd prefer.

(That would also remove the reference to `org.http4s.internal.CharPredicate.AlphaNum` on line 163, which may be problematic since it appears to be an internal http4s API?)

**Comment by bpholt** [comment]
@kubukoz @Baccata I think this is ready for review (CLA aside), at least for the portions that are specific to AWS. I think we agreed in Discord that the signing changes don't need to be configurable from Smithy traits, so I didn't make any changes in that direction. 

I wasn't sure if the changes to `fromSmithy4sHttpUri` and `fromSmithy4sHttpRequest` should be propagated out or not. Right now, I think `@simpleRestJson` would not use the new encoding rules (although it probably should), because `SimpleRestJsonCodecs#makeClientCodecs` doesn't set `encodeReservedCharacters` to true, but requests going to AWS will use the new rules since `AwsClient` does set that flag.

File: modules/aws-http4s/test/src-jvm/smithy4s/aws/internals/AwsSignatureTest.scala [hunk 2]
```diff
@@ -89,12 +89,77 @@ object AwsSignatureTest extends SimpleIOSuite with Checkers {
...
+         |  - smithy4sRegion: $smithy4sRegion
+         |  - awsRequest: ${awsRequest.show}
+         |""".stripMargin
+    }
 
-    val genAwsRequest = for {
+    val genPathSegment: Gen[String] =
+      Gen
+        .frequency(
+          1 -> Gen.oneOf("..", "."),
```
Comment (bpholt) [inline modules/aws-http4s/test/src-jvm/smithy4s/aws/internals/AwsSignatureTest.scala:133]: Path normalization should remove these when it's enabled, so these dot segments are sometimes added to the path to make sure that happens.

File: modules/aws-http4s/test/src-jvm/smithy4s/aws/internals/AwsSignatureTest.scala [hunk 3]
```diff
@@ -89,12 +89,77 @@ object AwsSignatureTest extends SimpleIOSuite with Checkers {
...
+            .stringOf(
+              Gen.frequency(
+                9 -> Gen.oneOf(
+                  Gen.alphaNumChar,
+                  // this list from https://github.com/http4s/http4s/blob/ac5c5cc40f732df238ef38c8ef8f571b4e66bfe3/core/shared/src/main/scala/org/http4s/internal/UriCoding.scala#L26
+                  Gen.oneOf("-_.~".toIndexedSeq)
+                ),
+                1 -> Gen.oneOf(
+                  // this list from https://github.com/http4s/http4s/blob/ac5c5cc40f732df238ef38c8ef8f571b4e66bfe3/core/shared/src/main/scala/org/http4s/Uri.scala#L1069
+                  "!$&'()*+,;=:/?@".toIndexedSeq
```
Comment (bpholt) [inline modules/aws-http4s/test/src-jvm/smithy4s/aws/internals/AwsSignatureTest.scala:144]: These are characters that should get encoded if they're present in a path segment and encoding is enabled.

File: modules/aws-http4s/test/src-jvm/smithy4s/aws/internals/AwsSignatureTest.scala [hunk 4]
```diff
@@ -89,12 +89,77 @@ object AwsSignatureTest extends SimpleIOSuite with Checkers {
...
+            _,
+            toSkip = org.http4s.internal.CharPredicate.AlphaNum ++ "-_.~"
+          )
+        )
+
+    val genAwsRequest: Gen[SdkHttpFullRequest] = for {
       httpMethod <- Gen.oneOf(SdkHttpMethod.values().toList)
       host <- Gen.identifier
-      path <- Gen.listOfN(3, Gen.identifier).map(_.mkString("/"))
+      path <- Gen
```
Comment (bpholt) [inline modules/aws-http4s/test/src-jvm/smithy4s/aws/internals/AwsSignatureTest.scala:159]: When I was working on this and trying to address some edge cases, I commented out this generator and used a variety of generators specific to the edge case I was working on. For example, things like `Gen.const("foo/../.")`, `Gen.const("this:might:be:an:aws:arn")`, etc. Unfortunately I didn't save the specifics, but they should be encoded in this generator.

**Comment by bpholt** [comment]
Unfortunately I think this implementation escapes too much. After running the protocol compliance suite from Alloy and Smithy itself, I noticed that the spec actually says only values inserted into the URI [via the `httpLabel` trait](https://smithy.io/2.0/spec/http-bindings.html#id7) should actually be escaped in this way. Notably, literal values that were already in the URI should _not_ be escaped. 

(Or at least, that's how I'm now reading [the spec](https://smithy.io/2.0/spec/http-bindings.html#literal-character-sequences): "Patterns with no labels will match only requests containing the exact literal characters declared in the pattern", and then "Every label … MUST have the [`httpLabel` trait](https://smithy.io/2.0/spec/http-bindings.html#httplabel-trait)". Then, the serialization rules that describe escaping only apply to the `httpLabel` trait.)

I don't think we have enough information in `fromSmithy4sHttpUri` to make that distinction, so I'm taking a step back. My current thought is to try to change `smithy4s.http.internals.SchemaVisitorPathEncoder` and maybe `smithy4s.internals.SchemaVisitorPatternEncoder` to escape labelled values, using the escaping algorithm used by Smithy.

**Comment by bpholt** [comment]
FWIW, I think we need to merge disneystreaming/alloy#210 before this will be fully tested by the Alloy compliance tests, but when I ran that locally before rebasing, it passed. (For some reason I can't get it to run now after pulling the latest updates; I get errors like "Unable to resolve trait `aws.protocols#ec2QueryName`" when executing `http4s / Test / managedResources` and `http4s / Test / envVars`. Maybe I'm not publishing Alloy correctly locally? I'm not very familiar with Mill.)

**Comment by bpholt** [comment]
https://github.com/disneystreaming/smithy4s/blob/87897b8514c4e79074770197bd759f0b92971c8d/modules/aws-http4s/src/smithy4s/aws/AwsClient.scala#L91-L108

Other than `AWS_REST_JSON_1`, should these codec constructors use the HTTP label encoding changes I introduced, or should I leave the old behavior in place?

File: modules/aws-http4s/src/smithy4s/aws/internals/AwsSigning.scala (hunk 1 above)
Comment (Baccata) [reply modules/aws-http4s/src/smithy4s/aws/internals/AwsSigning.scala]: yeah let's move :+1:, thank you for raising the concern

File: modules/core/src/smithy4s/internals/StructurePatternRefinementProvider.scala [hunk 5]
```diff
@@ -67,7 +67,7 @@ object StructurePatternRefinementProvider {
           )
         )
     (input: A) => {
-      val result = encoder.encode(input)
+      val result = encoder.encode(input, false) // TODO should this encode?
```
Comment (Baccata) [inline modules/core/src/smithy4s/internals/StructurePatternRefinementProvider.scala]: > // TODO should this encode? 

No need

File: modules/core/src/smithy4s/http/internals/PathEncode.scala [hunk 6]
```diff
@@ -19,12 +19,14 @@ package smithy4s.http.internals
 import smithy4s.capability.Contravariant
 
 trait PathEncode[A] { self =>
-  def encode(a: A): List[String]
-  def encodeGreedy(a: A): List[String]
+  def encode(a: A, urlEncode: Boolean): List[String]
+  def encodeGreedy(a: A, urlEncode: Boolean): List[String]
```
Comment (Baccata) [inline modules/core/src/smithy4s/http/internals/PathEncode.scala]: I'm not sure whether it makes sense to do it at this level to be honest. I'd rather it was a constructor parameter (in the `SchemaVisitorPathEncode` construct, for instance).

File: modules/core/src/smithy4s/http/internals/PathEncode.scala [hunk 7]
```diff
@@ -55,4 +65,56 @@ object PathEncode {
...
+
+        case ((acc, Vector('%', '2')), 'F') if ignoreSlashes =>
+          (acc :+ '/') -> Vector.empty
+
+        case ((acc, Vector('%', '2')), 'F') if !ignoreSlashes =>
+          acc ++ Vector('%', '2', 'F') -> Vector.empty
+
+        case ((acc, lookback), c) =>
+          (acc ++ lookback :+ c) -> Vector.empty
+      }
```
Comment (Baccata) [inline modules/core/src/smithy4s/http/internals/PathEncode.scala]: Would you mind rewriting it imperatively using a `Vector.builder` as an accumulator ? (Or even a StringBuilder)

File: modules/core/src/smithy4s/http/internals/SchemaVisitorPathEncoder.scala [hunk 8]
```diff
@@ -78,7 +78,7 @@ object SchemaVisitorPathEncoder
       fields: Vector[Field[S, _]],
       make: IndexedSeq[Any] => S
   ): MaybePathEncode[S] = {
-    type Writer = S => List[String]
+    type Writer = (S, Boolean) => List[String]
```
Comment (Baccata) [inline modules/core/src/smithy4s/http/internals/SchemaVisitorPathEncoder.scala]: See https://github.com/disneystreaming/smithy4s/pull/1668/files#r2039441673

File: modules/http4s/test/src/smithy4s/http4s/SimpleRestJsonComplianceSuite.scala [hunk 9]
```diff
@@ -95,7 +95,8 @@ object SimpleRestJsonComplianceSuite extends ProtocolComplianceSuite {
       val baseUri = uri"http://localhost"
       val suppliedHost =
         testHost.map(host => Uri.unsafeFromString(s"http://$host"))
-      SimpleRestJsonBuilder(service)
+      SimpleRestJsonBuilder
+        .apply(service)
```
Comment (Baccata) [inline modules/http4s/test/src/smithy4s/http4s/SimpleRestJsonComplianceSuite.scala]: is that a syntactic preference or did something change in a source-incompatible manner ?

**Comment by Baccata** [comment]
> Other than AWS_REST_JSON_1, should these codec constructors use the HTTP label encoding changes I introduced, or should I leave the old behavior in place? 

Let's leave it unchanged for now and revisit on a case-by-case basis (unless we find some evidence by analysis the spec that some protocols are inherently and unilateraly abiding by one rule)

**Comment by Baccata** [comment]
A few comments, but great work overall !

File: modules/http4s/test/src/smithy4s/http4s/SimpleRestJsonComplianceSuite.scala (hunk 9 above)
Comment (bpholt) [reply modules/http4s/test/src/smithy4s/http4s/SimpleRestJsonComplianceSuite.scala]: I think I probably added it temporarily so I could more easily click through to the implementation, and then forgot to revert it. I'll switch it back.

File: modules/core/src/smithy4s/http/internals/PathEncode.scala (hunk 6 above)
Comment (bpholt) [reply modules/core/src/smithy4s/http/internals/PathEncode.scala]: I'm not sure how to implement it at the instance level and also [maintain the distinction between the old `path` and the new `encodedPath` methods in the implementation of `HttpEndpoint[I]`](https://github.com/Dwolla/smithy4s/blob/a0aad90f8b0b5b19e50417448b8aeb5467558a69/modules/core/src/smithy4s/http/HttpEndpoint.scala#L75-L76). 

I guess I could change `SchemaVisitorPathEncoder` from an object to a class, and then construct two `PathEncoder[I]`s in `HttpEndpoint.cast`, where `HttpEndpoint[I]`s are constructed, in case downstream code calls the old `path` method? Is that what you meant?

In addition to not disturbing the public API (most of this work has been to internal APIs, but `HttpEndpoint` is not marked as internal), we also need to maintain the distinction in behavior in order to continue the existing behavior in the AWS codecs other than `AWS_REST_JSON_1`, [as requested](https://github.com/disneystreaming/smithy4s/pull/1668#issuecomment-2796740816).

**Comment by bpholt** [comment]
> > Other than AWS_REST_JSON_1, should these codec constructors use the HTTP label encoding changes I introduced, or should I leave the old behavior in place?
> 
> Let's leave it unchanged for now and revisit on a case-by-case basis (unless we find some evidence by analysis the spec that some protocols are inherently and unilateraly abiding by one rule)

FYI, it looks like `AWS_REST_XML` also requires encoding; without it, some of the compliance tests fail.

File: modules/core/src/smithy4s/http/internals/PathEncode.scala (hunk 7 above)
Comment (bpholt) [reply modules/core/src/smithy4s/http/internals/PathEncode.scala]: Addressed in 0c3df5ec7.

File: modules/core/src/smithy4s/http/internals/PathEncode.scala (hunk 6 above)
Comment (bpholt) [reply modules/core/src/smithy4s/http/internals/PathEncode.scala]: Addressed in fd284bd75.

**Comment by Baccata** [comment]
Labelled as "DO NOT MERGE" while the CLA situation is being worked out

File: modules/core/src/smithy4s/http/internals/PathEncode.scala (hunk 6 above)
Comment (Baccata) [reply modules/core/src/smithy4s/http/internals/PathEncode.scala]: Thank you, that is indeed what I meant :+1:

File: modules/aws-http4s/test/src-jvm/smithy4s/aws/internals/AwsSignatureTest.scala [hunk 10]
```diff
@@ -89,12 +89,77 @@ object AwsSignatureTest extends SimpleIOSuite with Checkers {
       awsRequest: SdkHttpFullRequest
   )
   object TestInput {
-    implicit val show: Show[TestInput] = Show.fromToString
+    implicit val showContentStreamProvider: Show[ContentStreamProvider] =
+      Show.show { csp =>
+        scala.io.Source.fromInputStream(csp.newStream()).mkString
+      }
+
+    implicit val showSdkHttpFullRequest: Show[SdkHttpFullRequest] = Show.show {
```
Comment (Baccata) [inline modules/aws-http4s/test/src-jvm/smithy4s/aws/internals/AwsSignatureTest.scala:97]: this is much appreciated, thank you !

**Comment by kubukoz** [comment]
ima put this back in draft instead of the scary label :P

**Review by Baccata — APPROVED** [review]
Happy with this in its current state. 

@bpholt has provided the signed corporate-CLA on behalf of Dwolla, which makes him exempt from having to sign the individual-CLA

@kubukoz, want to give it pass ?

File: modules/core/src/smithy4s/http/HttpEndpoint.scala [hunk 11]
```diff
@@ -23,8 +23,14 @@ import smithy4s.http.internals.SchemaVisitorPathEncoder
 
 trait HttpEndpoint[I] {
   // Returns a list of path segments that should be appended to the base URL. These are not URL-encoded.
+  @deprecated(
+    "Path segments should be encoded in a specific way per the Smithy spec, so use encodedPath instead",
+    "v0.18.34"
```
Comment (kubukoz) [inline modules/core/src/smithy4s/http/HttpEndpoint.scala]: it's gonna be .40 now, and maybe without the `v`

File: modules/core/src/smithy4s/http/HttpEndpoint.scala [hunk 12]
```diff
@@ -61,15 +67,23 @@ object HttpEndpoint {
           )
         )
 
-      encoder <- SchemaVisitorPathEncoder(
+      labelEncodingEncoder <- new SchemaVisitorPathEncoder(true)(
```
Comment (kubukoz) [inline modules/core/src/smithy4s/http/HttpEndpoint.scala]: suggestion: let's use a named param for the boolean

File: modules/core/src/smithy4s/http/HttpEndpoint.scala (hunk 11 above)
Comment (kubukoz) [reply modules/core/src/smithy4s/http/HttpEndpoint.scala]: also, should we actually deprecate it, if we keep using `path` in the case of `if rawHttpLabelValues`? I suppose the plan is to remove `rawHttpLabelValues` later and make `true` the default?

**Review by kubukoz — COMMENTED** [review]
please add an entry in the changelog file, it'll be 0.18.40 at this point

File: modules/core/src/smithy4s/http/HttpEndpoint.scala (hunk 11 above)
Comment (bpholt) [reply modules/core/src/smithy4s/http/HttpEndpoint.scala]: Yes, I think that's the plan.

File: modules/core/src/smithy4s/http/HttpEndpoint.scala [hunk 13]
```diff
@@ -24,8 +24,14 @@ import scala.annotation.tailrec
 
 trait HttpEndpoint[I] {
   // Returns a list of path segments that should be appended to the base URL. These are not URL-encoded.
+  @deprecated(
+    "Path segments should be encoded in a specific way per the Smithy spec, so use encodedPath instead",
+    "0.18.40"
+  )
   def path(input: I): List[String]
 
+  def encodedPath(input: I): List[String]
```
Comment (bpholt) [inline modules/core/src/smithy4s/http/HttpEndpoint.scala:33]: I'm a little surprised this new method didn't trigger a MiMa error, but I didn't see anything that would have suppressed it, so I'm hoping it's ok.

File: build.sbt [hunk 14]
```diff
@@ -286,11 +286,19 @@ lazy val core = projectMatrix
...
+        "smithy4s.http.HttpUnaryClientCodecs#HttpUnaryClientCodecsBuilderImpl.*"
+      ),
       // Breaking bin-compat to walk back ambiguous methods introduced in
       // https://github.com/disneystreaming/smithy4s/pull/1669
       ProblemFilters.exclude[IncompatibleMethTypeProblem](
         "smithy4s.http.HttpUnaryServerRouter.partialFunction"
-      )
+      ),
+      // originating in an Alloy update that removed ProtoCompactOffsetDateTime
+      ProblemFilters.exclude[MissingClassProblem]("alloy.proto.ProtoCompactOffsetDateTime"),
```
Comment (bpholt) [inline build.sbt:302]: These came in with the Alloy update. I'm not sure if filtering them out was the right call or not.

Comment (kubukoz) [reply build.sbt:302]: oh, I see now. The trait is no longer being generated...

looks good.

File: modules/core/src/smithy4s/http/HttpEndpoint.scala (hunk 13 above)
Comment (kubukoz) [reply modules/core/src/smithy4s/http/HttpEndpoint.scala:33]: I don't get it either, tbh. At least the hope is that nobody creates HttpEndpoint manually (as in, not via its `unapply` and `cast`)

btw. maybe we should seal this trait.

Comment (kubukoz) [reply modules/core/src/smithy4s/http/HttpEndpoint.scala:33]: I tried a similar case and it emits `ReversedMissingMethod`. We actually have a filter for all of these (`ProblemFilters.exclude[ReversedMissingMethodProblem]("smithy4s.*")` in `project/MimaVersionPlugin.scala`) because it's a forward-compatibility problem.

personally I don't buy it, it's a binary risk like any other. What saves us is that nobody else should be implementing `HttpEndpoint`, as per the comment above.

File: project/Dependencies.scala [hunk 15]
```diff
@@ -31,7 +31,7 @@ object Dependencies {
 
   val Alloy = new {
     val org = "com.disneystreaming.alloy"
-    val alloyVersion = "0.3.26"
+    val alloyVersion = "0.3.28-1-4a6c61"
```
Comment (kubukoz) [inline project/Dependencies.scala]: **issue:** just noting to ourselves to use a stable number before we merge

File: modules/core/src/smithy4s/http/HttpEndpoint.scala (hunk 13 above)
Comment (bpholt) [reply modules/core/src/smithy4s/http/HttpEndpoint.scala:33]: Yep, makes sense. I missed `MimaVersionPlugin`.

> personally I don't buy it, it's a binary risk like any other. What saves us is that nobody else should be implementing HttpEndpoint, as per the comment above.

Are you suggesting I change something so that it wouldn't emit a `ReversedMissingMethod` error if they weren't being suppressed? I'm not sure what the best way to do that is; none of the options I can think of are very good. (e.g. use `???`, or delegate to `path`)

Comment (kubukoz) [reply modules/core/src/smithy4s/http/HttpEndpoint.scala:33]: I wasn't suggesting any particular action in this case, just a more general wondering...ment for the other maintainers (@Baccata I'm looking at you) :)

in this case I think we're pretty safe because in practice this is "kinda sealed"

Comment (bpholt) [reply modules/core/src/smithy4s/http/HttpEndpoint.scala:33]: Ok. I made the changes necessary to seal `HttpEndpoint`, but I'm happy to revert them if that's not the path we want to take.

Comment (kubukoz) [reply modules/core/src/smithy4s/http/HttpEndpoint.scala:33]: oh I wasn't suggesting to actually seal it yet.

I would probably keep it open - the reasoning is, if someone has a custom implementation, their experience will be:

1. runtime linkage error
2. they add the missing implementation, whatever it is
3. no more error

whereas if we seal it now, they may be unable to build with the new smithy4s at all.

Not super realistic, but who knows. I'd leave this kind of change for 0.19

Comment (bpholt) [reply modules/core/src/smithy4s/http/HttpEndpoint.scala:33]: Ok, works for me. I unsealed it.

**Review by kubukoz — APPROVED** [review]
just waiting for https://github.com/disneystreaming/alloy/actions/runs/16455931297 now so that we can bump alloy here

File: project/Dependencies.scala (hunk 15 above)
Comment (bpholt) [reply project/Dependencies.scala]: ✅

**Review by kubukoz — APPROVED** [review]
